stablehlo/dialect/Base.h

```diff
@@ -51,6 +57,16 @@ bool isCompatibleForHloTypeInference(Type tp1, Type tp2);
 // sparsity.
 bool isCompatibleForHloTypeInference(TypeRange tp1, TypeRange tp2);
 
+// Infer single most specific return type from inputTypes with support for
+// bounds. (Size, bound) of each dimension of the return type will be inferred
+// from corresponding dimensions of every inputType by merging them. If
```

Comment:
In case of concatenate dim, add the requirement that all types must be ranked? Also, add invariance check for that as well.

Reply:
in `inferMostSpecificType()` if all the types are unranked, it will return `inputTypes[0]`.  So far I think it's enough even for concat dim. The callers can pass in ranked or unranked as well with no worries.